src/poly/commitment.rs

```diff
@@ -1,245 +1,40 @@
-use super::{
-    query::{ProverQuery, VerifierQuery},
-    strategy::Guard,
-    Coeff, LagrangeCoeff, Polynomial,
-};
-use crate::poly::Error;
-use crate::transcript::{EncodedChallenge, TranscriptRead, TranscriptWrite};
-use ff::Field;
-use halo2curves::CurveAffine;
-use rand_core::RngCore;
-use std::{
-    fmt::Debug,
-    io::{self},
-    ops::{Add, AddAssign, Mul, MulAssign},
-};
-
-/// Defines components of a commitment scheme.
-pub trait CommitmentScheme {
-    /// Application field of this commitment scheme
-    type Scalar: Field;
-
-    /// Elliptic curve used to commit the application and witnesses
-    type Curve: CurveAffine<ScalarExt = Self::Scalar>;
-
-    /// Constant prover parameters
-    type ParamsProver: for<'params> ParamsProver<
-        'params,
-        Self::Curve,
-        ParamsVerifier = Self::ParamsVerifier,
-    >;
-
-    /// Constant verifier parameters
-    type ParamsVerifier: for<'params> ParamsVerifier<'params, Self::Curve>;
-
-    /// Wrapper for parameter generator
-    fn new_params(k: u32) -> Self::ParamsProver;
-
-    /// Wrapper for parameter reader
-    fn read_params<R: io::Read>(reader: &mut R) -> io::Result<Self::ParamsProver>;
-}
-
-/// Parameters for circuit sysnthesis and prover parameters.
-pub trait Params<'params, C: CurveAffine>: Sized + Clone {
-    /// Multi scalar multiplication engine
-    type MSM: MSM<C> + 'params;
-
-    /// Logaritmic size of the circuit
-    fn k(&self) -> u32;
-
-    /// Size of the circuit
-    fn n(&self) -> u64;
-
-    /// Downsize `Params` with smaller `k`.
-    fn downsize(&mut self, k: u32);
-
-    /// Generates an empty multiscalar multiplication struct using the
-    /// appropriate params.
-    fn empty_msm(&'params self) -> Self::MSM;
-
-    /// This commits to a polynomial using its evaluations over the $2^k$ size
-    /// evaluation domain. The commitment will be blinded by the blinding factor
-    /// `r`.
-    fn commit_lagrange(
+//! Trait for a commitment scheme
+use crate::poly::{Blind, Coeff, Error, Polynomial, ProverQuery, VerifierQuery};
+use crate::transcript::{Hashable, Sampleable, Transcript};
+use ff::PrimeField;
+use halo2curves::serde::SerdeObject;
+
+/// Public interface for a Polynomial Commitment Scheme (PCS)
+pub trait PolynomialCommitmentScheme<F: PrimeField>: Clone {
+    /// Parameters needed to generate a proof in the PCS
+    type Parameters;
+    /// Parameters needed to verify a proof in the PCS
+    type VerifierParameters: From<Self::Parameters>;
+    /// Type of a committed polynomial
+    type Commitment: Clone + Copy + PartialEq + SerdeObject + Send + Sync; // Thinking of having an MSM here
+
+    /// Setup the parameters for the PCS
+    fn setup(k: u32) -> Self::Parameters;
+    /// Create a new instantiation of the PCS
+    fn new(params: Self::Parameters) -> Self;
```

Comment:
I would leave it for now, but it could be better to pass the `Parameters` to `commit` and `open` instead of passing `self`. Then removing this `new` function.

Reply:
I'll open an issue to address it in a different PR. https://github.com/input-output-hk/halo2/issues/20